File: sdk/python/term_sdk/agent.py
"""
Base Agent class for Term Challenge.
"""

from abc import ABC, abstractmethod
from .types import Request, Response


class Agent(ABC):
    """
    Base class for Term Challenge agents.
    
    Implement `solve()` to create your agent:
    
    ```python
    from term_sdk import Agent, Request, Response, run
    
    class MyAgent(Agent):
        def solve(self, req: Request) -> Response:
            # Your logic here
            if req.step == 1:
                return Response.cmd("ls -la")
            return Response.done()
    
    if __name__ == "__main__":
        run(MyAgent())
    ```
    
    Lifecycle:
        1. `setup()` - Called once before processing (optional)
        2. `solve()` - Called for each step (required)
        3. `cleanup()` - Called after task completes (optional)
    """
    
    def setup(self) -> None:
        """
        Initialize resources before processing.
        
        Override to set up LLM clients, load data, etc.
        
        Example:
            ```python
            def setup(self):
                self.llm = LLM(model="gpt-4o")
                self.history = []
            ```
        """
        pass
    
    @abstractmethod
    def solve(self, request: Request) -> Response:
        """
        Process a request and return a response.
        
        This is called for each step of the task.
        
        Args:
            request: Contains instruction, step, output, etc.
        
        Returns:
            Response with command to execute or task_complete=True
        
        Example:
            ```python
            def solve(self, req: Request) -> Response:
                # First step: explore
                if req.first:
                    return Response.cmd("ls -la")
                
                # Check for errors
                if req.failed:
                    return Response.cmd("pwd")
                
                # Check output
                if req.has("hello", "world"):
                    return Response.done()
                
                # Default action
                return Response.cmd("echo 'working...'")
            ```
        """
        raise NotImplementedError
    
    def cleanup(self) -> None:
        """
        Clean up resources after task completes.
        
        Override to close connections, save state, etc.
        
        Example:
            ```python
            def cleanup(self):
                self.llm.close()
            ```
        """
        pass

File: sdk/rust/src/agent.rs
//! Agent trait for Term Challenge.

use crate::{Request, Response};

/// Trait for Term Challenge agents.
///
/// Implement `solve()` to create your agent:
///
/// ```rust,no_run
/// use term_sdk::{Agent, Request, Response};
///
/// struct MyAgent;
///
/// impl Agent for MyAgent {
///     fn solve(&mut self, req: &Request) -> Response {
///         if req.step == 1 {
///             return Response::cmd("ls -la");
///         }
///         Response::done()
///     }
/// }
/// ```
pub trait Agent {
    /// Initialize the agent (optional).
    fn setup(&mut self) {}
    
    /// Process a request and return a response.
    fn solve(&mut self, request: &Request) -> Response;
    
    /// Clean up resources (optional).
    fn cleanup(&mut self) {}
}
